include/StateChecker.php

```diff
@@ -0,0 +1,322 @@
+     * 
+     * @param string $key
+     * @return boolean
+     */
+    protected function isDetailedKey($key)
+    {
+        $detailedKey = preg_match('/\w+\:\:/', $key);
+        return $detailedKey;
+    }
+    
+    /**
+     * 
+     * @param string $hash
+     * @param string $key
+     * @return boolean
+     */
+    protected function checkHash($hash, $key)
+    {
+        $detailedKey = $this->isDetailedKey($key);
+        $needToStore = !$detailedKey || ($detailedKey && StateCheckerConfig::get('storeDetails'));
+        
+        if (!isset($this->hashes[$key])) {
+            if ($needToStore) {
+                $this->hashes[$key] = $hash;
+            }
+        }
+        if ($needToStore) {
+            $match = $this->hashes[$key] == $hash;
+        } else {
+            $match = true;
+        }
+        return $match;
+    }
+    
+    /**
+     * 
+     * @param mixed $data should be serializable
+     * @param string $key
+     * @return string
+     * @throws StateCheckerException
+     */
+    protected function getHash($data, $key)
+    {
+        if (!$serialized = serialize($data)) {
+            throw new StateCheckerException('Serialize object failure');
+        }
+        $hash = md5($serialized);
+        
+        if (!$this->checkHash($hash, $key)) {
+            throw new StateCheckerException('Hash doesn\'t match at key "' . $key . '".');
+        }
+        
+        if (StateCheckerConfig::get('saveTraces')) {
+            $this->traces[$key][] = debug_backtrace();
+        }
+        
+        return $hash;
+    }
+    
+    // ------------ DATABASE ----------------
+    
+    /**
+     * 
+     * @param mysqli_result $resource
+     * @return array
+     */
+    protected function getMysqliResoults(mysqli_result $resource)
+    {
+        $rows = [];
+        while ($row = $resource->fetch_assoc()) {
+            $rows[] = $row;
+        }
+        return $rows;
+    }
+    
+    /**
+     * 
+     * @return array
+     * @throws StateCheckerException
+     */
+    protected function getDatabaseTables()
+    {
+        if (!$tables = $this->db->tablesLike('')) {
+            throw new StateCheckerException('get tables failure');
+        }
+        return $tables;
+    }
+    
+    /**
+     * 
+     * @return string
+     */
+    protected function getDatabaseHash()
+    {
+        $tables = $this->getDatabaseTables();
+        $hashes = [];
+        foreach ($tables as $table) {
+            $rows = $this->getMysqliResoults($this->db->query('SELECT * FROM ' . $table));
+            $hashes[] = $this->getHash($rows, 'database::' . $table);
+        }
+        $hash = $this->getHash($hashes, 'database');
+        return $hash;
+    }
+    
+    // ------------- FILE SYSTEM ---------------
+    
+    /**
+     * 
+     * @param string $path
+     * @return array
+     * @throws StateCheckerException
+     */
+    protected function getFiles($path = '.')
+    {
+        if (!$realpath = realpath($path)) {
```

Comment:
Assigning or comparison? Please extract it to avoid confusion.

Reply:
Assign. one `=` means assign.

Reply:
ah, ok sorry I get what you mean.